File: Python/39-平衡二叉树.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def IsBalanced_Solution(self, root):
        # write code here
        if not root:
            return True
        if abs(self.TreeDepth(root.left) - self.TreeDepth(root.right)) <= 1:
            return self.IsBalanced_Solution(root.left) and self.IsBalanced_Solution(root.right)
        return False
        
    def TreeDepth(self, root):
        if not root:
            return 0
        return max(self.TreeDepth(root.left), self.TreeDepth(root.right)) + 1

File: C++/39-平衡二叉树.cpp
class Solution
{
public:
    bool flag;
    bool IsBalanced_Solution(TreeNode *pRoot)
    {
        flag = true;
        TreeDepth(pRoot);
        return flag;
    }

    int max(int a, int b)
    {
        return a >= b ? a : b;
    }

    int TreeDepth(TreeNode *root)
    {
        if (!root)
            return 0;
        int left = TreeDepth(root->left);
        int right = TreeDepth(root->right);
        if (abs(left - right) > 1)
        {
            flag = false;
        }
        return 1 + max(left, right);
    }
};